### PR TITLE
Convert utf8 string to GameStrings using current font mapping + add string viewer to rwviewer

### PR DESCRIPTION
**Pull request by madebr**
# What is this pr trying to solve?
`GameString`s are essentially a vector of indices in the font texture (=`GameStringChar`)
These font textures contain special characters, such as ß,æ,№, ...
Some of these characters have a utf-8 encoding of more then 1 byte.
some logic is needed to convert a `std::string` to a `GameString` correctly.
This pr solves this by using the conversion utf-8 <-> unicode <-> GameStringChar.

Also, not all textures contain the same character at the same location.
Therefore, converting unicode points to `GameStringChar`s, requires knowledge of the font texture used.

# Overview of the patches
- moved all string data related things to rwlib
-`rwlib/source/fonts/Unicode.hpp/cpp` contains functions/classes to convert utf-8 to unicode
- added all unicode points used by gta3 to `rwlib/source/fonts/Unicode.hpp`
- `rwlib/source/fonts/FontMapsGta3.hpp/cpp` contains the maps from unicode <-> `GameStringChar`
- added tests for utf8 <-> unicode
- `rwlib/soruce/fonts/GameTexts.hpp/cpp` needs to know of the used font to convert utf8 <-> `GameStringChar`
- I changed all font references of 0 to `FONT_PAGER`, 1 to `FONT_PRICEDOWN` and 2 to `FONT_ARIAL`.
- rwviewer can now display all texts of the language files and can display random text. This should help/verify #129

rwviewer screenshot:
screenshot from 2018-06-23 04-00-41

### CODE REVIEW

File: rwengine/src/engine/GameData.hpp [hunk 1]
```diff
@@ -18,7 +18,7 @@
 #include <rw/forward.hpp>
 
 #include <data/AnimGroup.hpp>
-#include <data/GameTexts.hpp>
+#include <fonts/GameTexts.hpp>
```
Comment (ShFil119) [inline rwengine/src/engine/GameData.hpp]: order

File: rwengine/src/engine/GameState.hpp [hunk 2]
```diff
@@ -9,7 +9,7 @@
 #include <glm/glm.hpp>
 #include <glm/gtc/quaternion.hpp>
 
-#include <data/GameTexts.hpp>
+#include <fonts/GameTexts.hpp>
```
Comment (ShFil119) [inline rwengine/src/engine/GameState.hpp]: order

File: rwviewer/views/TextViewer.cpp [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+        emit gameStringChanged(gameString);
+    } catch (std::out_of_range &) {
+        return;
+    }
+}
+
+TextViewer::TextViewer(QWidget* parent, Qt::WindowFlags f)
+    : ViewerInterface(parent, f) {
+
+    auto dataLayout = new QVBoxLayout;
```
Comment (ShFil119) [inline rwviewer/views/TextViewer.cpp]: Here and below I'd use value initialization.

File: rwlib/source/fonts/FontMapGta3.hpp [hunk 4]
```diff
@@ -0,0 +1,26 @@
+#ifndef _RWLIB_FONTS_FONTMAPGTA3_HPP_
```
Comment (danhedron) [inline rwlib/source/fonts/FontMapGta3.hpp]: I think this file is more at home inside engine, as it contains a lot of hard-coded game specific stuff.

File: rwlib/source/fonts/Unicode.hpp [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+#define _RWLIB_FONTS_UNICODE_HPP_
+
+#include <cstdint>
+#include <iosfwd>
+#include <sstream>
+
+/**
+ * unicode_t represent a unicode data point. (UTF-32)
+ */
+using unicode_t = std::uint32_t;
```
Comment (danhedron) [inline rwlib/source/fonts/Unicode.hpp]: Can we not use char32_t? these are just UTF-32 encoded characters right?

File: rwviewer/views/TextViewer.cpp (hunk 3 above)
Comment (madebr) [reply rwviewer/views/TextViewer.cpp]: I know what you mean, but this is "the qt way".
All sample code of qt is written this way.

File: rwlib/source/fonts/FontMapGta3.hpp (hunk 4 above)
Comment (madebr) [reply rwlib/source/fonts/FontMapGta3.hpp]: After discussion on irc, it can stay in rwlib (for now).

File: rwlib/source/fonts/Unicode.hpp (hunk 5 above)
Comment (madebr) [reply rwlib/source/fonts/Unicode.hpp]: Yup we can and we will.

**Comment by madebr** [comment]
- Use `UNICODE_BLACK_LEFT_POINTING_POINTER` instead of `UNICODE_BLACK_LEFT_TRIANGLE`
- Use `UNICODE_BLACK_RIGHT_POINTING_POINTER` instead of `UNICODE_BLACK_RIGHT_TRIANGLE`
- Add ability to iterate over all mappings of a `FontMap`.

File: .appveyor.yml [hunk 6]
```diff
@@ -40,7 +40,8 @@ init:
 install:
   - cd "%APPVEYOR_BUILD_FOLDER%"
   - for /f %%i in ('git rev-parse HEAD') do set COMMIT_HASH=%%i
-  - if NOT "%APPVEYOR_REPO_COMMIT%" == "%COMMIT_HASH%" echo "Appveyor hash does not match checkout out hash" && exit 1
+  - echo APPVEYOR_REPO_COMMIT=%APPVEYOR_REPO_COMMIT%, COMMIT_HASH=%COMMIT_HASH%
+  - if NOT "%APPVEYOR_REPO_COMMIT%" == "%COMMIT_HASH%" echo "Appveyor git hash does not match git checkout hash"
```
Comment (danhedron) [inline .appveyor.yml]: These changes are already merged

File: rwlib/source/fonts/Unicode.hpp [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+    /**
+     * @brief iss This utf8 stream holds the wrapped string and returns bytes to convert to unicode.
+     */
+    std::istringstream iss;
+};
+
+/**
+ * Unicode data points used by OpenRW.
+     */
+    enum UnicodeValue {
```
Comment (ShFil119) [inline rwlib/source/fonts/Unicode.hpp]: enum class?

File: rwviewer/models/TextModel.cpp [hunk 8]
```diff
@@ -0,0 +1,76 @@
+#include "TextModel.hpp"
+#include <iostream>
+
+#include <QBrush>
+
+TextModel::TextModel(QObject *parent)
+     : QAbstractTableModel(parent), m_font(FONT_PAGER) {
```
Comment (ShFil119) [inline rwviewer/models/TextModel.cpp]: m_font can be initialized using default member initializer.

File: rwviewer/views/TextViewer.cpp [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+        std::transform(filename.begin(), filename.end(), filename.begin(), ::tolower);
+        if (!filename.compare("text")) {
+            textPath = p;
+            break;
+        }
+    }
+    if (!textPath.string().length()) {
+        throw std::runtime_error("text directory not found in gamedata path");
+    }
+    std::vector<std::string> names;
```
Comment (ShFil119) [inline rwviewer/views/TextViewer.cpp]: It should be doable to use `reserve`.

File: rwviewer/views/TextViewer.hpp [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    GameString currentGameString;
+    font_t currentFont;
+    int currentFontSize;
+
+    void updateRender();
+
+    void setGameString(const GameString &gameString);
+    std::vector<std::string> getFontTextureNames();
+public:
+    TextViewer(QWidget* parent = 0, Qt::WindowFlags f = 0);
```
Comment (ShFil119) [inline rwviewer/views/TextViewer.hpp]: `nullptr`?

File: rwengine/src/render/TextRenderer.hpp [hunk 11]
```diff
@@ -30,10 +30,10 @@ class TextRenderer {
      * @todo Can this be merged with the gamestate text entries?
      */
     struct TextInfo {
-        enum TextAlignemnt { Left = 0, Right = 1, Center = 2 };
+        enum TextAlignment { Left = 0, Right = 1, Center = 2 };
```
Comment (ShFil119) [inline rwengine/src/render/TextRenderer.hpp]: It also can be enum class.

File: rwlib/source/fonts/GameTexts.hpp [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+ * Keys are small so should be subject to SSO
+ */
+using GameStringKey = std::string;
+
+/**
+ * Index to used font.
+ */
+using font_t = size_t;
+
+static const font_t FONT_PAGER = 0;
```
Comment (ShFil119) [inline rwlib/source/fonts/GameTexts.hpp]: Here and below even can be `constexpr`.

File: rwlib/source/fonts/Unicode.cpp [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+        c[1] = 0x80 | ((unicode >> 12) & 0x3f);
+        c[2] = 0x80 | ((unicode >> 6) & 0x3f);
+        c[3] = 0x80 | (unicode & 0x3f);
+        return 4;
+    } else {
+        return unicode_to_utf8(UNICODE_REPLACEMENT_CHARACTER, c);
+    }
+}
+
+Utf8UnicodeIterator::Utf8UnicodeIterator() : m_is(nullptr), m_finished(true) {
```
Comment (ShFil119) [inline rwlib/source/fonts/Unicode.cpp]: default member initializer

File: rwviewer/ViewerWidget.cpp [hunk 14]
```diff
@@ -32,6 +35,7 @@ ViewCamera OrbitCamera (const glm::vec2& viewPort, const glm::vec2& viewAngles,
 ViewerWidget::ViewerWidget(QOpenGLContext* context, QWindow* parent)
     : QWindow(parent)
     , context(context)
+    , textInfos()
```
Comment (ShFil119) [inline rwviewer/ViewerWidget.cpp]: Is it even needed?
// below can be done using default member initializer

File: rwlib/source/fonts/Unicode.hpp (hunk 7 above)
Comment (madebr) [reply rwlib/source/fonts/Unicode.hpp]: Enum classes would make UnicodeValue strongly-typed. Implicitly converting a UnicodeValue to a unicode_t becomes impossible and requires a static_cast. https://en.wikipedia.org/wiki/C%2B%2B11#Strongly_typed_enumerations

File: rwviewer/views/TextViewer.cpp (hunk 9 above)
Comment (madebr) [reply rwviewer/views/TextViewer.cpp]: That vector just contains a list of all text localisations (american, english, french, ...) . Normally it is short so reserve should not be needed.

File: rwlib/source/fonts/Unicode.cpp [hunk 15]
```diff
@@ -22,11 +22,11 @@ size_t unicode_to_utf8(unicode_t unicode, char c[4]) {
         c[3] = 0x80 | (unicode & 0x3f);
         return 4;
     } else {
-        return unicode_to_utf8(UNICODE_REPLACEMENT_CHARACTER, c);
+        return unicode_to_utf8(UnicodeValue::UNICODE_REPLACEMENT_CHARACTER, c);
     }
 }
 
-Utf8UnicodeIterator::Utf8UnicodeIterator() : m_is(nullptr), m_finished(true) {
+Utf8UnicodeIterator::Utf8UnicodeIterator() {
```
Comment (ShFil119) [inline rwlib/source/fonts/Unicode.cpp]: `= default` and can be moved to header. ;)

File: rwviewer/ViewerWidget.hpp [hunk 16]
```diff
@@ -78,23 +84,24 @@ public slots:
     GameWorld* _world = nullptr;
     GameRenderer* _renderer = nullptr;
 
+    std::vector<TextRenderer::TextInfo> textInfos = {};
```
Comment (ShFil119) [inline rwviewer/ViewerWidget.hpp]: `= {}`is unneeded

**Comment by danhedron** [comment]
lgtm

**Pull request merged**